### ISSUE TITLE
XGrammar support for Spec V2

### ISSUE THREAD

**Issue by JustinTong0323** [sgl-project/sglang#13019]
Required when `tool_choice="required"`

**Comment by jiapingW** [comment]
> [@hnyls2002](https://github.com/hnyls2002) [@JustinTong0323](https://github.com/JustinTong0323) need to look deeper but i assume it's because there's no grammar backend reference in v2 file https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/speculative/eagle_info_v2.py unlike v1 file https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/speculative/eagle_info.py#L9

I debugged the program and found that since Spec v2 does not introduce any grammar-related operations, it does not perform the relevant processing. The `batch.spec_info.grammar` is None.

**Comment by pdasgup** [comment]
thanks @jiapingW . will you be raising the fix for this?

**Comment by hnyls2002** [comment]
Adapt these two logics from worker v1 to worker v2

- https://github.com/sgl-project/sglang/blob/85b8c5c4cde2b3da7d63fba10e657b79b11484ab/python/sglang/srt/speculative/eagle_worker.py#L670-L687
- https://github.com/sgl-project/sglang/blob/85b8c5c4cde2b3da7d63fba10e657b79b11484ab/python/sglang/srt/speculative/eagle_worker.py#L653-L658

**Comment by jiapingW** [comment]
I'm modifying its logic. In v1, the batch's `spec_info` included grammar information, but in v2, it's not being retrieved here: https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/speculative/eagle_worker_v2.py#L587. It seems this parameter wasn't initialized in a previous initialization step.

**Comment by jiapingW** [comment]
I implemented a similar solution using the patch below, which does use grammar and handle logits, but issues such as duplicate output still exist and I'm still working on it.
```python

diff --git a/python/sglang/srt/speculative/eagle_info_v2.py b/python/sglang/srt/speculative/eagle_info_v2.py
index 8d334a460..cd886e18a 100644
--- a/python/sglang/srt/speculative/eagle_info_v2.py
+++ b/python/sglang/srt/speculative/eagle_info_v2.py
@@ -1,13 +1,14 @@
 from __future__ import annotations
 
 from dataclasses import dataclass
-from typing import TYPE_CHECKING, Any
+from typing import TYPE_CHECKING, Any, Optional
 
 import torch
 import torch.nn.functional as F
 import triton
 import triton.language as tl
 
+from sglang.srt.constrained.base_grammar_backend import BaseGrammarObject
 from sglang.srt.layers.logits_processor import LogitsProcessorOutput
 from sglang.srt.managers.schedule_batch import ModelWorkerBatch, ScheduleBatch
 from sglang.srt.mem_cache.common import (
@@ -238,11 +239,14 @@ class EagleVerifyInputV2Mixin:
         self: EagleVerifyInput,
         batch: ModelWorkerBatch,
         logits_output: LogitsProcessorOutput,
+        vocab_mask: Optional[torch.Tensor] = None,  # For grammar
     ):
         """
         Verify and find accepted tokens based on logits output and batch
         (which contains spec decoding information).
         """
+        from .forkedpdb import ForkedPdb
+        ForkedPdb().set_trace()
         bs = len(batch.seq_lens)
         sampling_info = batch.sampling_info
         next_token_logits = logits_output.next_token_logits
@@ -257,6 +261,13 @@ class EagleVerifyInputV2Mixin:
         )
         accept_length = torch.empty((bs,), dtype=torch.int32, device=device)
 
+        # Apply grammar mask
+        if vocab_mask is not None:
+            assert self.grammar is not None
+            self.grammar.apply_vocab_mask(
+                logits=logits_output.next_token_logits, vocab_mask=vocab_mask
+            )
+
         # Sample tokens
         if sampling_info.is_all_greedy or _is_npu:
             target_predict = torch.argmax(next_token_logits, dim=-1)

diff --git a/python/sglang/srt/speculative/eagle_worker_v2.py b/python/sglang/srt/speculative/eagle_worker_v2.py
index 6e7bc5573..308c084c9 100644
--- a/python/sglang/srt/speculative/eagle_worker_v2.py
+++ b/python/sglang/srt/speculative/eagle_worker_v2.py
@@ -33,6 +33,7 @@ from sglang.srt.speculative.spec_utils import (
     detect_nan,
     draft_tp_context,
     load_token_map,
+    generate_token_bitmask_v2,
 )
 from sglang.srt.utils.common import (
     empty_context,
@@ -294,7 +295,8 @@ class EagleDraftWorker(BaseDraftWorker):
             tree_mask_buf,
             position_buf,
         )
-
+        from .forkedpdb import ForkedPdb
+        # ForkedPdb().set_trace()
         return EagleVerifyInput(
             draft_token=draft_tokens,
             custom_mask=tree_mask,
@@ -551,6 +553,8 @@ class EAGLEWorkerV2(BaseSpecWorker):
         pass
 
     def forward_batch_generation(self, model_worker_batch: ModelWorkerBatch):
+        from .forkedpdb import ForkedPdb
+        # ForkedPdb().set_trace()
         if model_worker_batch.forward_mode.is_decode():
             draft_input: EagleDraftInput = model_worker_batch.spec_info
             assert draft_input.is_draft_input()
@@ -581,6 +585,8 @@ class EAGLEWorkerV2(BaseSpecWorker):
         batch: ModelWorkerBatch,
         cur_allocate_lens: torch.Tensor,
     ):
+        from .forkedpdb import ForkedPdb
+        ForkedPdb().set_trace()
         # Since batch.seq_lens is allocated in another stream, we need
         # record_stream() to prevent pytorch gc and reuse the gpu memory
         # while forward_stream is still running.
@@ -628,6 +634,36 @@ class EAGLEWorkerV2(BaseSpecWorker):
         )
         logits_output = forward_batch_output.logits_output
 
+        # grammar usage
+        if batch.sampling_info.grammars is not None:
+            retrieve_next_token_cpu = batch.spec_info.retrive_next_token.cpu()
+            retrieve_next_sibling_cpu = batch.spec_info.retrive_next_sibling.cpu()
+            draft_tokens_cpu = batch.spec_info.draft_token.view(
+                batch.spec_info.retrive_next_token.shape
+            ).cpu()
+
+        vocab_mask = None
+        if batch.sampling_info.grammars is not None:
+            # Generate the logit mask for structured output.
+            # Overlap the CPU operations for bitmask generation with the forward pass.
+            # ForkedPdb().set_trace()
+            vocab_mask = generate_token_bitmask_v2(
+                batch,
+                batch.spec_info,
+                retrieve_next_token_cpu,
+                retrieve_next_sibling_cpu,
+                draft_tokens_cpu,
+                batch.sampling_info.vocab_size,
+            )
+
+            if vocab_mask is not None:
+                assert batch.sampling_info.grammars is not None
+                vocab_mask = vocab_mask.to(batch.spec_info.retrive_next_token.device)
+                # NOTE (sk): otherwise, this vocab mask will be the one from the previous extend stage
+                # and will be applied to produce wrong results
+                batch.sampling_info.vocab_mask = None
+
+
         # Sample
         if self.enable_nan_detection:
             detect_nan(logits_output)
@@ -635,7 +671,7 @@ class EAGLEWorkerV2(BaseSpecWorker):
             predict,
             accept_length,
             accept_index,
-        ) = verify_input.sample(batch, logits_output)
+        ) = verify_input.sample(batch, logits_output, vocab_mask)
         new_seq_lens = batch.seq_lens + accept_length
         verify_done = torch.get_device_module(self.device).Event()
         verify_done.record()
diff --git a/python/sglang/srt/speculative/spec_utils.py b/python/sglang/srt/speculative/spec_utils.py
index ed949902d..67c582320 100644
--- a/python/sglang/srt/speculative/spec_utils.py
+++ b/python/sglang/srt/speculative/spec_utils.py
@@ -11,6 +11,7 @@ import triton
 import triton.language as tl
 from huggingface_hub import snapshot_download
 
+from sglang.srt.managers.schedule_batch import ModelWorkerBatch
 from sglang.srt.constrained.base_grammar_backend import BaseGrammarObject
 from sglang.srt.distributed.parallel_state import (
     GroupCoordinator,
@@ -608,6 +609,8 @@ def generate_token_bitmask(
     draft_tokens_cpu: torch.Tensor,
     vocab_size: int,
 ):
+    from .forkedpdb import ForkedPdb
+    ForkedPdb().set_trace()
     """
     Generate the logit mask for structured output.
     Draft model's token can be either valid or invalid with respect to the grammar.
@@ -651,6 +654,60 @@ def generate_token_bitmask(
     return allocate_token_bitmask
 
 
+
+def generate_token_bitmask_v2(
+    batch: ModelWorkerBatch,
+    verify_input: EagleVerifyInput,
+    retrieve_next_token_cpu: torch.Tensor,
+    retrieve_next_sibling_cpu: torch.Tensor,
+    draft_tokens_cpu: torch.Tensor,
+    vocab_size: int,
+):
+    from .forkedpdb import ForkedPdb
+    # ForkedPdb().set_trace()
+    """
+    Generate the logit mask for structured output.
+    Draft model's token can be either valid or invalid with respect to the grammar.
+    We need to perform DFS to
+    1. figure out which tokens are accepted by the grammar.
+    2. if so, what is the corresponding logit mask.
+    """
+
+    num_draft_tokens = draft_tokens_cpu.shape[-1]
+
+    allocate_token_bitmask = None
+    grammar = None
+    for i in range(retrieve_next_token_cpu.shape[0]):
+        if batch.sampling_info.grammars[i] is not None:
+            if allocate_token_bitmask is None:
+                allocate_token_bitmask = batch.sampling_info.grammars[i].allocate_vocab_mask(
+                    vocab_size=vocab_size,
+                    batch_size=draft_tokens_cpu.numel(),
+                    device="cpu",
+                )
+            grammar = batch.sampling_info.grammars[i]
+            s = time.perf_counter()
+            traverse_tree(
+                retrieve_next_token_cpu[i],
+                retrieve_next_sibling_cpu[i],
+                draft_tokens_cpu[i],
+                batch.sampling_info.grammars[i],
+                allocate_token_bitmask[
+                    i * num_draft_tokens : (i + 1) * num_draft_tokens
+                ],
+            )
+            tree_traverse_time = time.perf_counter() - s
+            if tree_traverse_time > TREE_TRAVERSE_TIME_THRESHOLD:
+                logger.warning(
+                    f"Bit mask generation took {tree_traverse_time} seconds with "
+                    f"grammar: {batch.sampling_info.grammars[i]}"
+                )
+
+    verify_input.grammar = grammar
+    return allocate_token_bitmask
+
+
+
 def load_token_map(token_map_path: str) -> List[int]:
     if not os.path.exists(token_map_path):
         cache_dir = snapshot_download(

```
Result:
```python
--- Validating Results ---
Result 1/3: 'uusseseseseseseseseseseseserseseserserserserserserser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@'
  -> FAIL: Text does not match regex pattern '^user@example\.com$'.
Result 2/3: 'useuserrrrrrrrr@rr@r@r@r@r@r@r@r@r@r@r@r@r@r@r@r@r@r@r@r@r@r@r@r@r@r@r@r@r@r@r@r@r@er@r@r@r@r@er@er@er@er@er@er@er@er@er@er@er@er@er'
  -> FAIL: Text does not match regex pattern '^user@example\.com$'.
Result 3/3: 'uusseseseseseseseseseseseserseseserserserserserserser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@ser@'
  -> FAIL: Text does not match regex pattern '^user@example\.com$'.
```

**Comment by jiapingW** [comment]
I think the following code is also need to align: https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/speculative/eagle_info.py#L371
```python
        for i, (req, accept_index_row) in enumerate(zip(batch.reqs, accept_index_cpu)):
            for j, idx in enumerate(accept_index_row):
                if idx == -1:
                    break
                id = predict_cpu[idx]
                req.output_ids.append(id)
                req.check_finished()
                if req.finished():
                    has_finished = True
                    # set all tokens after finished token to -1 and break
                    accept_index[i, j + 1 :] = -1
                    break
                else:
                    if req.grammar is not None:
                        try:
                            req.grammar.accept_token(id)
                        except ValueError as e:
                            logger.info(
                                f"{i=}, {req=}\n" f"{accept_index=}\n" f"{predict=}\n"
                            )
                            raise e
```

**Comment by pdasgup** [comment]
@Ubospica @hnyls2002 

we are still observing some structured output issues with specv2. below is an example for qwen235b with v0.5.6. we observed this for other models as well.

```
root@predictor-resource-pool-1401806956670222336-5ffbd455b5-f2sq6:/sgl-workspace/sglang# curl http://localhost:7080/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{"model":"your_model_name","messages":[{"content":"Your task is to extract the name of the person from the following text.","role":"system"},{"content":"I'\''m John Doe.","role":"user"}],"response_format":{"json_schema":{"name":"extract_name","schema":{"properties":{"name":{"type":"string"}},"required":["name"],"title":"ExtractName","type":"object"}},"type":"json_schema"},"max_tokens":1000,"stream":false}'


{"id":"aaec92f5334a44ff85bc381f7fcb1180","object":"chat.completion","created":1765403485,"model":"your_model_name","choices":[{"index":0,"message":{"role":"assistant","content":"{\"name\": \"John!! Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe} Doe}}}} Doe!!} Doe Doe Doe Doe Doe Doe Doe} Doe! Doe! Doe!! Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe! Doe!! Doe!!! Doe! Doe!! Doe. Doe. Doe!!! Doe! Doe!! Doe! Doe!!! Doe!oeoeoe Doe Doeoe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doeoe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe Doe John! John Doe Doe!! Doe!!oeoeoeoe Doeoeoe Doe Doeoe Doe Doe!oe Doeoeoeoe Doe Doeoeoe Doeoeoe Doe Doeoeoe Doe Doe Doeoeoe Doeoeoeoe Doeoe Doeoe!!oe!!oeoeoe!!oeoeoe!!oe!oe!oe!oe! Doe!! Doe!!oe!!oeoeoe!oe! Doe!! Doe!!oe!! Doe!!oeoe Doeoeoeoe Doeoeoeoeoe Doe Doeoe Doe Doeoeoe!oe!oe!!oe! Doe!! Doe!!oe!!oe!!oe!!oe!oe! Doe!! Doe!!oe!oe!oe!!oe!oe!oe!! Doe!! Doe!!oe!oe! Doe!! Doe!!oe!oe!!oe! Doe!! Doe!!oe!!oe!oe!!oe Doeoeoe Doe Doeoeoe Doe Doeoeoe Doe Doe Doe Doeoe Doe! Doe!oe!oe!oe!oe! Doe!! Doe!!oe!!!oe!oe!!! Doe!!oe! Doe!!oe Doeoe Doe Doe Doeoe Doe Doe!oe Doeoe Doeoe!!oe!! Doe!!oe! Doe!!oeoeoeoe Doeoe Doeoe Doeoeoeoeoeoe Doeoeoeoe! Doe!oe! Doe!!oeoeoe Doeoe!oe! Doe!! Doe!!oe!oe Doeoe Doe Doe!oe Doeoe Doe!!oe!oe!!oe!!oe!! Doe!!oe!oe!oe!!! Doe.oe!!oe! Doe!!oe!oe!! Doe!!oe!oe!oe! Doe!!oe!oe!oe!! Doe Doe Doe Doe Doe Doe! Doe!!!!oe Doe Doeoeoeoe Doe! Doe!oe.oe!!oe.!oeoe Doeoe Doeoe!oe!oe!.oe!.oe Doeoe Doeoeoeoeoeoeoe!oe!oe!!oe!.oe!!oe!oe!!oeoe Doeoeoeoe Doeoe Doe Doeoe Doe Doe!oe!!oe!!oe.!oe!!oe.oe!oe.oe!!oe!oe!!oe.oe!oe!!oe. Doe!!oe!!oe!.oe!oe!!oe.oe!.oe!!oe!.oe!!oe!oe.!oe!oeoeoeoe Doeoe Doeoeoe Doeoe Doe Doe!oe!.oe!!oe!!oe!oeoeoe Doeoeoeoeoe!! Doe.oe!!oe!!oe.!oe!!oe.!oe.!oe.!oe.!oe.!oe.!oe.!oe.!oe.!oe!!oe!.oe.!oeoe Doeoeoe!!!oe!.!oe!oe!oe!oe!.oe!.oe!.! Doe!!oe Doe Doeoeoeoeoeoeoeoe!oeoeoeoeoeoe!!oe!oe.!oe!oe.!oe Doe Doe Doeoe Doe!oe!! Doe!!oe.!oeoeoe Doeoe Doeoeoeoeoe Doe Doe Doeoe!!oe!!oe!oe!oe Doe Doe Doeoeoeoeoeoeoeoeoeoeoeoe!!oe!!oe!oe!oeoeoeoeoe!oe!oe.!!oe!!oe.oe.!oeoe Doeoe Doe!!!!oe!!oe!oe!oe!.=\"}","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":"stop","matched_stop":151645}],"usage":{"prompt_tokens":33,"total_tokens":972,"completion_tokens":939,"prompt_tokens_details":{"cached_tokens":3},"reasoning_tokens":0},"metadata":{"weight_version":"default"}}
```
